File: python3/maximum_points_you_can_obtain_from_cards.py
# Space   : O(1)
# Time    : O(k)

class Solution:
    def maxScore(self, cardPoints: List[int], k: int) -> int:
        n = len(cardPoints)
        ans = window = sum(cardPoints[:k])
        l, r = k-1, n-1

        while l >= 0:
            window += cardPoints[r] - cardPoints[l]
            ans = max(ans, window)
            l -= 1
            r -= 1

        return ans

File: golang/maximum_points_you_can_obtain_from_cards.go
package golang

// Space   : O(1)
// Time    : O(k)

func maxScore(cardPoints []int, k int) int {
	n := len(cardPoints)

	ans, total := 0, 0
	for j := 0; j < k; j++ {
		ans += cardPoints[j]
	}
	total = ans

	for i := k - 1; i >= 0; i-- {
		total += cardPoints[n-k+i] - cardPoints[i]
		ans = Max(ans, total)
	}

	return ans
}
